### PR TITLE
Enable Reclaiming of Escrowed Payments

### LINKED ISSUE #21: User SBAT reclaim funds associated with an unused invite after a certain period
### Expected Behavior

Users might send an invite to an user who decides not to take up the invite or may send an invite to an invalid phone number and the invite itself is never received. In these cases, after a certain period,  if an invite has not been activated, an inviter should be able to recover (most of) the escrow amount that they sent with that invite.

### Current Behavior

Once an invite is sent, the associated fees and escrow amount can only be redeemed by the invitee. If the invitee never redeems the invite, then the funds are lost in the void.

### PR DESCRIPTION
**Pull request by medhakothari**
### Description

Allows for the reclaiming of escrowed payments. Once the payment has expired, it gives the sender a notification on the wallet to either remind the sender through a text message, or allows the sender to reclaim the payment. 

### Tested

Tested the following tests between two accounts.

- [x] Alice sends an escrowed payment to Bob. Payment hasn't expired yet, so notification does not show up to Alice 
- [x]  Alice sends an escrowed payment to Bob. Payment expires, and Alice gets a notification on the home screen of wallet app. Alice is able to send a message or reclaim the payment successfully. 
- [x]  Alice sends an escrowed payment to Bob. Bob withdraws the payment. The payment expires, but as it was already withdrawn, the notification does not show up to Alice. 
- [x]  Alice sends an escrowed payment to Bob. The payment expires, and Alice reclaims the payment. Bob attempts to withdraw the payment, and is able to successfully redeem the invite, but **not** the escrowed payment

Screenshots of the notification and the Reclaim Payment Confirmation Screen:
Screen Shot 2019-07-25 at 3 07 03 PM   Screen Shot 2019-07-25 at 3 07 53 PM

### Other changes

- I added additional testing for the escrow smart contract

### Related issues

- Fixes #21

### Backwards compatibility

none

### CODE REVIEW

File: packages/mobile/src/escrow/actions.ts [hunk 1]
```diff
@@ -13,17 +20,21 @@ export interface EscrowedPayment {
   expirySeconds: BigNumber
 }
 
-// The number of seconds before the sender can revoke the payment.
-export const EXPIRY_SECONDS = 432000 // 5 days in seconds
+// The number of seconds before the sender can reclaim the payment.
+export const EXPIRY_SECONDS = 1
```
Comment (medhakothari) [inline packages/mobile/src/escrow/actions.ts]: I changed the default payment expiry time in the wallet from 5 days to 1 second so senders can instantaneously reclaim payments. Would appreciate thoughts/opinions!

Comment (jmrossy) [reply packages/mobile/src/escrow/actions.ts]: seems ok to me

File: packages/mobile/src/escrow/actions.ts [hunk 2]
```diff
@@ -13,17 +20,21 @@ export interface EscrowedPayment {
...
   TRANSFER_PAYMENT = 'ESCROW/TRANSFER_PAYMENT',
   RECLAIM_PAYMENT = 'ESCROW/RECLAIM_PAYMENT',
   GET_SENT_PAYMENTS = 'ESCROW/GET_SENT_PAYMENTS',
   STORE_SENT_PAYMENTS = 'ESCROW/STORE_SENT_PAYMENTS',
   RESEND_PAYMENT = 'ESCROW/RESEND_PAYMENT',
+  FETCH_RECLAIM_TRANSACTION_FEE = 'ESCROW/FETCH_RECLAIM_TRANSACTION_FEE',
+  SET_RECLAIM_TRANSACTION_FEE = 'ESCROW/SET_RECLAIM_TRANSACTION_FEE',
 }
 
+const TAG = 'escrow/actions'
```
Comment (jmrossy) [inline packages/mobile/src/escrow/actions.ts]: Nit: Move to top of file, just below imports

File: packages/mobile/src/escrow/actions.ts [hunk 3]
```diff
@@ -88,3 +105,35 @@ export const resendPayment = (paymentId: string): ResendPaymentAction => ({
   type: Actions.RESEND_PAYMENT,
   paymentId,
 })
+
+export const setReclaimTransactionFee = (suggestedFee: string): SetReclaimTransactionFeeAction => ({
+  type: Actions.SET_RECLAIM_TRANSACTION_FEE,
+  suggestedFee,
+})
+
+export const fetchReclaimSuggestedFee = (paymentID: string) => async (
```
Comment (jmrossy) [inline packages/mobile/src/escrow/actions.ts]: New async actions should all use redux saga instead of redux thunk. Please convert this to saga.

File: packages/mobile/src/escrow/actions.ts [hunk 4]
```diff
@@ -88,3 +105,35 @@ export const resendPayment = (paymentId: string): ResendPaymentAction => ({
...
+  try {
+    Logger.debug(`${TAG}/fetchSuggestedFee`, `Fetching reclaim transaction fee`)
+    const escrow = await getEscrowContract(web3)
+    const stableToken = await getStableTokenContract(web3)
+    const account = currentAccountSelector(getState())
+
+    const mockReclaimTx = await escrow.methods.revoke(paymentID)
+
+    const mockTxParams: any = { from: account, gasCurrency: stableToken._address }
+    const gas: BigNumber = new BigNumber(await mockReclaimTx.estimateGas(mockTxParams))
```
Comment (jmrossy) [inline packages/mobile/src/escrow/actions.ts]: Nit: No need for the explicit setting of type (`: BigNumber`) in the next 3 lines
In general, the only time we should have to do that is for the result of a `yield` call. Hopefully even that case will be fixed soon :)

File: packages/mobile/src/notifications/EscrowedPaymentReminderNotification.tsx [hunk 5]
```diff
@@ -54,6 +59,7 @@ export class EscrowedPaymentReminderNotification extends React.PureComponent<Pro
         onPress: () => {
           const reclaimPaymentInput = payment
           CeloAnalytics.track(CustomEventNames.clicked_escrowed_payment_notification)
+          this.props.fetchReclaimSuggestedFee(payment.paymentID)
```
Comment (jmrossy) [inline packages/mobile/src/notifications/EscrowedPaymentReminderNotification.tsx]: This could take a few seconds to load. Does the confirmation card show a spinner in the meantime? Or what does it show?

File: packages/mobile/src/escrow/saga.ts [hunk 6]
```diff
@@ -165,15 +167,19 @@ function* getSentPayments() {
...
         senderAddress: payment[1],
-        recipient: recipientsPhoneNumbers[sentPaymentIDs[i]],
+        recipientPhone: recipientPhoneNumber,
         paymentID: sentPaymentIDs[i],
         currency: SHORT_CURRENCIES.DOLLAR, // Only dollars can be escrowed
         amount: payment[3],
         timestamp: payment[6],
         expirySeconds: payment[7],
       }
+      if (recipientPhoneNumberToContact[recipientPhoneNumber]) {
```
Comment (jmrossy) [inline packages/mobile/src/escrow/saga.ts]: This approach is totally fine but just a suggestion for another way to handle this would be to add this line to the object definition just above:
`recipientContact: recipientPhoneNumberToContact[recipientPhoneNumber] || undefined`

File: packages/mobile/test/values.ts [hunk 7]
```diff
@@ -51,6 +51,15 @@ export const mockRecipientWithPhoneNumber: RecipientWithMobileNumber = {
   e164PhoneNumber: mockE164Number,
 }
 
+export const mockRecipientWithContact: RecipientWithContact = {
```
Comment (jmrossy) [inline packages/mobile/test/values.ts]: Why add this instead of using the existing `mockRecipient` above?

**Review by jmrossy — COMMENTED** [review]
Good progress! Some comment below :)

File: packages/mobile/src/escrow/actions.ts (hunk 1 above)
Comment (medhakothari) [reply packages/mobile/src/escrow/actions.ts]: @nityas thoughts?

File: packages/mobile/test/values.ts (hunk 7 above)
Comment (medhakothari) [reply packages/mobile/test/values.ts]: good catch, didn't see that one!

File: packages/mobile/src/notifications/EscrowedPaymentReminderNotification.tsx (hunk 5 above)
Comment (medhakothari) [reply packages/mobile/src/notifications/EscrowedPaymentReminderNotification.tsx]: Since this function isn't making any real transactions, I've noticed it usually takes less than a second. However, should I add a spinner here just in case? If so, should I follow the SUCCESS/FAILURE action format which was discussed on the slack today?

Comment (jmrossy) [reply packages/mobile/src/notifications/EscrowedPaymentReminderNotification.tsx]: In that case, I'm okay to leave this as is for now

File: packages/mobile/src/escrow/reducer.ts [hunk 8]
```diff
@@ -1,11 +1,14 @@
...
+import { RootState } from 'src/redux/reducers'
 
 export interface State {
   sentEscrowedPayments: EscrowedPayment[]
+  suggestedFee: string
 }
 
 export const initialState = {
   sentEscrowedPayments: [],
+  suggestedFee: '',
```
Comment (jmrossy) [inline packages/mobile/src/escrow/reducer.ts]: We try to avoid using empty strings, better to use `null` to indicate something doesn't have a value yet. This will require setting the type to `string | null` above

File: packages/mobile/src/escrow/saga.ts [hunk 9]
```diff
@@ -183,6 +190,30 @@ function* getSentPayments() {
...
+    const gas = new BigNumber(yield mockReclaimTx.estimateGas(mockTxParams))
+    const gasPrice = new BigNumber(yield call(fetchGasPrice))
+    const suggestedFeeInWei = gas.multipliedBy(gasPrice)
+
+    yield put(setReclaimTransactionFee(suggestedFeeInWei.toString()))
+
+    Logger.debug(TAG + '@fetchReclaimSuggestedFee', `New reclaim tx fee is: ${suggestedFeeInWei}`)
+  } catch (e) {
+    Logger.error(TAG + '@fetchReclaimSuggestedFee', 'Error fetching reclaim transaction fee', e)
+    throw e
```
Comment (jmrossy) [inline packages/mobile/src/escrow/saga.ts]: what catches this? Our sagas shouldn't propagate errors. Instead you should use `showError` to show a banner here

**Review by jmrossy — COMMENTED** [review]
Two small things, otherwise LGTM

File: packages/mobile/src/escrow/saga.ts [hunk 10]
```diff
@@ -191,6 +200,30 @@ function* getSentPayments() {
...
+    const gas = new BigNumber(yield mockReclaimTx.estimateGas(mockTxParams))
+    const gasPrice = new BigNumber(yield call(fetchGasPrice))
+    const suggestedFeeInWei = gas.multipliedBy(gasPrice)
+
+    yield put(setReclaimTransactionFee(suggestedFeeInWei.toString()))
+
+    Logger.debug(TAG + '@fetchReclaimSuggestedFee', `New reclaim tx fee is: ${suggestedFeeInWei}`)
+  } catch (e) {
+    yield put(showError(ErrorMessages.FETCH_RECLAIM_FEE_FAILED, ERROR_BANNER_DURATION))
+    throw e
```
Comment (jmrossy) [inline packages/mobile/src/escrow/saga.ts]: You resolved my comment but this throw is still here

File: packages/mobile/src/escrow/ReclaimPaymentConfirmationScreen.tsx [hunk 11]
```diff
@@ -97,6 +98,13 @@ class ReclaimPaymentConfirmationScreen extends React.Component<Props> {
   render() {
     const { t, fee } = this.props
     const payment = this.getReclaimPaymentInput()
+    const convertedAmount = divideByWei(payment.amount.toString())
+    if (!fee) {
+      Logger.error(TAG, 'Error fetching suggested reclaim fee, show error message')
```
Comment (jmrossy) [inline packages/mobile/src/escrow/ReclaimPaymentConfirmationScreen.tsx]: Render functions should be pure (no side-effects). 
We need to remove this logger call and error dispatch. Why is this here? Can the sagas that set fee handle this error setting? And what should this render if fee isn't set?

File: packages/mobile/src/escrow/saga.ts (hunk 10 above)
Comment (medhakothari) [reply packages/mobile/src/escrow/saga.ts]: whoops sorry about that!

File: packages/mobile/src/escrow/actions.ts (hunk 1 above)
Comment (medhakothari) [reply packages/mobile/src/escrow/actions.ts]: Decision was to keep it at 5 days -- changed back.

File: packages/mobile/src/escrow/ReclaimPaymentConfirmationScreen.tsx [hunk 12]
```diff
@@ -94,28 +103,41 @@ class ReclaimPaymentConfirmationScreen extends React.Component<Props> {
...
+    return this.props.isReclaiming ? (
+      <ActivityIndicator size="large" color={colors.celoGreen} />
+    ) : null
+  }
+
   render() {
-    const { t, fee } = this.props
+    const { t, isReclaiming, appConnected } = this.props
     const payment = this.getReclaimPaymentInput()
+    const convertedAmount = divideByWei(payment.amount.toString())
```
Comment (jmrossy) [inline packages/mobile/src/escrow/ReclaimPaymentConfirmationScreen.tsx]: Why do we need a toString() here?
It'd be more clever to make divideByWei accept other types, including null values :)

File: packages/mobile/src/escrow/ReclaimPaymentConfirmationScreen.tsx [hunk 13]
```diff
@@ -94,28 +103,41 @@ class ReclaimPaymentConfirmationScreen extends React.Component<Props> {
...
+    const convertedFee = divideByWei(this.getFee().toString())
```
Comment (jmrossy) [inline packages/mobile/src/escrow/ReclaimPaymentConfirmationScreen.tsx]: Ditto here with getFee

File: packages/mobile/src/escrow/ReclaimPaymentConfirmationScreen.tsx [hunk 14]
```diff
@@ -94,28 +103,41 @@ class ReclaimPaymentConfirmationScreen extends React.Component<Props> {
...
-          modifyButton={{ action: this.onPressEdit, text: t('cancel'), disabled: false }}
+          modifyButton={{ action: this.onPressEdit, text: t('cancel'), disabled: isReclaiming }}
         >
           <ReclaimPaymentConfirmationCard
-            recipient={payment.recipient}
+            recipientPhone={payment.recipientPhone}
+            recipientContact={payment.recipientContact}
             comment={payment.message}
-            amount={payment.amount}
+            amount={new BigNumber(convertedAmount)}
```
Comment (jmrossy) [inline packages/mobile/src/escrow/ReclaimPaymentConfirmationScreen.tsx]: I noticed that every place we use divideByWei, we then parse the result into a BIgNumber again. Can you please change divideByWei to return a BigNumber?

**Review by jmrossy — COMMENTED** [review]
Almost there! Just some nits, we should be able to merge today.
Also, please include a screenshot here.

File: packages/mobile/src/escrow/ReclaimPaymentConfirmationScreen.tsx (hunk 12 above)
Comment (medhakothari) [reply packages/mobile/src/escrow/ReclaimPaymentConfirmationScreen.tsx]: it's bc `payment.amount` is a  `BigNumber` not bc its null 
`divideByWei` can't take BigNums

Now that I think if it, `BigNumber.valueOf()` might be better and more explicit than `toString()`. Will change it to that

File: packages/mobile/src/escrow/ReclaimPaymentConfirmationScreen.tsx (hunk 14 above)
Comment (medhakothari) [reply packages/mobile/src/escrow/ReclaimPaymentConfirmationScreen.tsx]: ooh interesting, will do

**Review by jmrossy — APPROVED** [review]
